module/mempool/stdmap/backdata/arraylinkedlist/entitiesLinkedList.go

```diff
@@ -0,0 +1,350 @@
+package arraylinkedlist
+
+import (
+	"math/rand"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type EjectionMode string
+
+const (
+	RandomEjection = EjectionMode("random-ejection")
+	LRUEjection    = EjectionMode("lru-ejection")
+)
+
+// number of entries we try to eject before giving up and ejecting LRU
+const maximumRandomTrials = 10
+
+// EIndex is data type representing an entity index in EntityDoubleLinkedList.
+type EIndex uint32
+
+// doubleLinkedListPointer represents a slice-based linked list pointer. Instead of pointing
+// to a memory address, this pointer points to a slice index.
+//
+// Note: an "undefined" (i.e., nil) notion for this doubleLinkedListPointer corresponds to the
+// value of uint32(0). Hence, legit "pointer" values start from uint32(1).
+// doubleLinkedListPointer also furnished with methods to convert a "pointer" value to
+// a slice index, and reverse.
+type doubleLinkedListPointer struct {
+	pointerValue uint32
+}
+
+// isUndefined returns true if this pointer is set to its zero value. An undefined
+// slice-based pointer is equivalent to a nil address-based one.
+func (d doubleLinkedListPointer) isUndefined() bool {
+	return d.pointerValue == uint32(0)
+}
+
+// setUndefined sets sliced-based pointer to its undefined (i.e., nil equivalent) value.
+func (d *doubleLinkedListPointer) setUndefined() {
+	d.pointerValue = uint32(0)
+}
+
+// sliceIndex returns the slice-index equivalent of the pointer.
+func (d doubleLinkedListPointer) sliceIndex() EIndex {
+	return EIndex(d.pointerValue) - 1
+}
+
+// setPointer converts the input slice-based index into a slice-based pointer and
+// sets the underlying pointer.
+func (d *doubleLinkedListPointer) setPointer(sliceIndex EIndex) {
+	d.pointerValue = uint32(sliceIndex + 1)
+}
+
+// doubleLinkedListNode represents a slice-based double linked list node that
+// consists of a next and previous pointer.
+type doubleLinkedListNode struct {
+	next doubleLinkedListPointer
+	prev doubleLinkedListPointer
```

Comment:
Any reason to use `prev` instead of e.g. `last`? Am considering `last` is 'cleaner' because it's also the same length at `next` but is not abbreviated like `prev` but has a similar meaning?

Reply:
The `next` and `prev` are the [textbook notions](https://www.google.ca/books/edition/Introduction_To_Algorithms/NLngYyWFl_YC?hl=en&gbpv=1&dq=doubly+linked+list&pg=PA204&printsec=frontcover) of doubly-linked lists that are consistent [across literature](https://en.wikipedia.org/wiki/Doubly_linked_list).

Another example is taken from [GeeksForGeeks](https://www.geeksforgeeks.org/doubly-linked-list/):
![image](https://user-images.githubusercontent.com/19204398/148277906-65aad858-18ee-46c5-897a-e6752beae47c.png)

Reply:
It's true but it doesn't make it better :-)

For example, wikipedia [1] calls the `head` and `tail` of the double linked list `first` and `last` which have their own potentially misleading properties, and also they are words with different lengths which is annoying in the source code :-) Most literature also uses `first` and `last` although I have seen `head` and `last` too! :-) It's more difficult to find `head` and `tail`... :-(

But up to you... I think `prev` as an abbreviation is unlikely to be confused with another word (which is good) and is the same length as `next` (which is good) but admittedly could be confused as `last` as in `tail`... It's just annoying that it's an abbreviation that is so commonly used and accepted :-)

But if you really care about following the common literature then you probably want to consider changing `head` and `tail` into `first` and `last` or if you can't stomach that (and it's difficult for me to have suggested it because `first` is longer than `last`... OCD coming out there :-) ) `head` and `last`...

[1] https://en.wikipedia.org/wiki/Doubly_linked_list